### ISSUE TITLE
Only keep in-flight jobs in memory

### ISSUE THREAD

**Issue by GoogleCodeExporter** [uzh/gc3pie#162]
```
1) What part of the model would need changes:

`Engine` in `core.py`.  Possibly also `TaskCollection` in `dag.py`.


2) What is the reason why the changes is proposed:

The Selectome experiment needs to run a *very* large job campaign for
their validation purposes, in the order of *million* of jobs.

That's not currently possible, and RAM usage is the main barrier:
running 35k jobs with `gamess` eats about 1GB of memory, so we can
estimate a memory occupation of ~25kB per `Application` object in
memory. 

This gives a limit of about 100k-150k jobs that we can handle with 4GB
of RAM.  Too low.


3) What is the proposal:

In `Engine`, we only need the "live" (SUBMITTED and RUNNING) jobs to be kept in
memory, since we are going to update them at every cycle.  Then we
would have a limit of 100k live jobs, which is 20x the size of SMSCG.

Jobs in state NEW and TERMINATED can reside on disk; we only need to
pull of a few NEW jobs from disk when submissions are attempted.

A similar proposal holds for the `.tasks` list within
`TaskCollection`, though the data structure might be different as we
don't sort jobs by state there.

In summary, we would need a Python data structure that:
  * keeps all data on permament storage (whatever format)
  * allows iteration; we don't need random access to a specific item
  * as iteration proceeds, moves objects from disk to memory in small chunks (e.g., one by one)
  * when iteration is stopped, moves back objects to disk and frees up memory
I don't know of any Python package that provides just this; we need 
to do some research.


```

Original issue reported on code.google.com by `riccardo.murri@gmail.com` on 22 Mar 2011 at 5:17
- Blocking: #213
